FirebaseAuth/Sources/Auth/FIRAuth.m

```diff
@@ -485,85 +491,124 @@ - (nullable instancetype)initWithAPIKey:(NSString *)APIKey
+#endif  // TARGET_OS_IOS || TARGET_OS_TV || TARGET_OS_MACCATALYST
+  // Continue with the rest of initialization in the work thread.
+  __weak FIRAuth *weakSelf = self;
+  dispatch_async(FIRAuthGlobalWorkQueue(), ^{
+    // Load current user from Keychain.
+    FIRAuth *strongSelf = weakSelf;
+    if (!strongSelf) {
+      return;
+    }
+    NSString *keychainServiceName =
+        [FIRAuth keychainServiceNameForAppName:strongSelf->_firebaseAppName];
+    if (keychainServiceName) {
+      strongSelf->_keychainServices =
+          [[FIRAuthKeychainServices alloc] initWithService:keychainServiceName];
+      strongSelf.storedUserManager =
+          [[FIRAuthStoredUserManager alloc] initWithServiceName:keychainServiceName];
+    }
+
+    NSError *error;
+    NSString *storedUserAccessGroup =
+        [strongSelf.storedUserManager getStoredUserAccessGroupWithError:&error];
+    if (!error) {
+      if (!storedUserAccessGroup) {
+        FIRUser *user;
+        if ([strongSelf getUser:&user error:&error]) {
+          strongSelf.tenantID = user.tenantID;
+          [strongSelf updateCurrentUser:user byForce:NO savingToDisk:NO error:&error];
+          self->_lastNotifiedUserToken = user.rawAccessToken;
         } else {
-          [strongSelf internalUseUserAccessGroup:storedUserAccessGroup error:&error];
-          if (error) {
-            FIRLogError(kFIRLoggerAuth, @"I-AUT000001",
-                        @"Error loading saved user when starting up: %@", error);
-          }
+          FIRLogError(kFIRLoggerAuth, @"I-AUT000001",
+                      @"Error loading saved user when starting up: %@", error);
         }
       } else {
-        FIRLogError(kFIRLoggerAuth, @"I-AUT000001",
-                    @"Error loading saved user when starting up: %@", error);
+        [strongSelf internalUseUserAccessGroup:storedUserAccessGroup error:&error];
+        if (error) {
+          FIRLogError(kFIRLoggerAuth, @"I-AUT000001",
+                      @"Error loading saved user when starting up: %@", error);
+        }
       }
+    } else {
+      FIRLogError(kFIRLoggerAuth, @"I-AUT000001", @"Error loading saved user when starting up: %@",
+                  error);
+    }
 
 #if TARGET_OS_IOS
-      // Initialize for phone number auth.
-      strongSelf->_tokenManager = [[FIRAuthAPNSTokenManager alloc] initWithApplication:application];
+    static Class applicationClass = nil;
+    // iOS App extensions should not call [UIApplication sharedApplication], even if UIApplication
+    // responds to it.
+    if (![GULAppEnvironmentUtil isAppExtension]) {
+      Class cls = NSClassFromString(@"UIApplication");
+      if (cls && [cls respondsToSelector:@selector(sharedApplication)]) {
+        applicationClass = cls;
+      }
+    }
+    UIApplication *application = [applicationClass sharedApplication];
+
+    // Initialize for phone number auth.
+    strongSelf->_tokenManager = [[FIRAuthAPNSTokenManager alloc] initWithApplication:application];
```

Comment:
@ryanwilson This line seems to be failing static analysis now (since `application` can be nil if it's an iOS app extension and `initWithApplication` requires a non-nil argument). However, this is not a behavior change from how it was before, from what I can tell, since all I did was move this logic around. Is there a reason that it would fail static analysis now, but not before?

Reply:
Hmm, I agree it doesn't seem like a behavior change but maybe the analyzer couldn't pick it up before for some reason. Are these properties that require `application` optional, and could we skip creating them if `application` is `nil`?

Reply:
Done.